gitprovider.go

```diff
@@ -0,0 +1,115 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gitprovider
+
+import (
+	"errors"
+)
+
+var (
+	// ErrProviderNoSupport describes that the provider doesn't support the requested feature
+	ErrProviderNoSupport = errors.New("no provider support for this feature")
+	// ErrDomainUnsupported describes the case where e.g. a Github provider used for trying to get
+	// information from e.g. gitlab.com
+	ErrDomainUnsupported = errors.New("the given client doesn't support handling requests for this domain")
+
+	// ErrUnauthorized describes that that the request was unauthorized by the Git provider server
+	ErrUnauthorized = errors.New("unauthorized request for this resource")
+	// ErrRateLimited is returned if a request hit the rate limit of the Git provider server
+	ErrRateLimited = errors.New("hit rate limit")
+
+	// ErrNotTopLevelOrganization describes the case where it's mandatory to specify a top-level organization
+	// (e.g. for accessing teams), but a sub-organization was passed as the OrganizationRef
+	ErrNotTopLevelOrganization = errors.New("the specified organization wasn't a top-level one as expected")
+
+	// ErrAlreadyExists is returned by .Create() requests if the given resource already exists.
+	// Use .Reconcile() instead if you want to idempotently create the resource
+	ErrAlreadyExists = errors.New("resource already exists, cannot create object. Use Reconcile() to idempotently create.")
+	// ErrNotFound is returned by .Get() and .Update() calls if the given resource doesn't exist
+	ErrNotFound = errors.New("the requested resource was not found")
+
+	// ErrFieldRequired specifies the case where a required field isn't populated at use time
+	ErrFieldRequired = errors.New("field is required")
+	// ErrFieldInvalid specifies the case where a field isn't populated in a valid manner
+	ErrFieldInvalid = errors.New("field is invalid")
+	// ErrFieldEnumInvalid specifies the case where the given value isn't part of the known values in the enum
+	ErrFieldEnumInvalid = errors.New("field value isn't known to this enum")
+
+	// ErrURLUnsupportedScheme is returned if an URL without the HTTPS scheme is parsed
+	ErrURLUnsupportedScheme = errors.New("unsupported URL scheme, only HTTPS supported")
+	// ErrURLUnsupportedParts is returned if an URL with fragment, query values and/or user information is parsed
+	ErrURLUnsupportedParts = errors.New("organization or repository URLs cannot have fragments, query values nor user information")
```

Comment:
I think we should either create dedicated errors for organization and repository or make the current messages generic as in: 

```
ErrURLUnsupportedParts = errors.New("URL cannot have fragments, query values nor user information")
```

Doesn't make sense to me to say `URLs` when in fact you validate a single URL, having `organization or repository URLs` confuses the end-user as it doesn't tell it what URL is actually wrong. I think the error should contain the URL string.

Reply:
Thanks. Changed to your suggestion, that's better. The actual URL which is wrong will be added as context at runtime, but this can error can be used for `errors.Is(err, ErrURLUnsupportedParts)`

Reply:
Great! thanks